examples/tests/benchmark_vllm.py

```diff
@@ -0,0 +1,315 @@
+            file_handle,
+            (
+                f"load {num_prompts} examples\n"
+                f"total time : {elapsed_time}\n"
+                f"average time per batch : {elapsed_time}"
+        ))
+
+        total_token = 0
+        max_len = 0
+        if outputs is None:
+            max_len = 0
+            for request in requests:
+                total_token += request.output_len * sample
+                max_len = max(max_len, request.output_len)
+        else:
+            for responses_for_one_prompt in outputs:
+                for response in responses_for_one_prompt:
+                    token_ids = response.token_ids
+                    total_token += len(token_ids)
+                    max_len = max(max_len, len(token_ids))
+        avg_len = total_token / (num_prompts * sample)
+
+        print_and_write_to_file(
+            file_handle,
+            (
+                f"total output token : {total_token}\n"
+                f"max output len : {max_len}\n"
+                f"avg output len : {avg_len}\n"
+                f"output token throuput per second : {total_token / elapsed_time}"
+        ))
+
+        if outputs is None:
+            return
+
+        print_and_write_to_file(
+            file_handle,
+            f"A total number of {len(outputs) * sample} outputs are generated"
+        )
+        s = ""
+        weird_count = 0
+        index = 0
+        for ex in outputs:
+            for iid, rsp in enumerate(ex):
+                rsp_text = rsp.text
+                s += f"rsp {index*32+iid} : {rsp_text}\n"
+                if "same same" in rsp_text:
+                    weird_count += 1
+            index += 1
+        print_and_write_to_file(file_handle, f"weird_count {weird_count}")
+        file_handle.write(s)
+
+def benchmark_vllm(args):
+    if args.vllm_use_ray_spmd_worker:
+        os.environ['VLLM_USE_RAY_SPMD_WORKER'] = '1'
+        os.environ['VLLM_USE_RAY_COMPILED_DAG'] = '1'
+        os.environ['VLLM_USE_RAY_COMPILED_DAG_NCCL_CHANNEL'] = '0'
+    else:
+        os.environ['VLLM_USE_RAY_SPMD_WORKER'] = '0'
+        os.environ['VLLM_USE_RAY_COMPILED_DAG'] = '0'
+    tp_size = args.tensor_parallel_size
+    pp_size = args.pipeline_parallel_size
+    enforce_eager = args.enforce_eager
+    model_name = args.model_name
+    model_path = args.model_path
+    tokenizer = AutoTokenizer.from_pretrained(model_path)
+    num_prompts = args.num_prompts
+
+    data_file = args.data
+    with open(data_file) as f:
+        raw_ex_batch = [json.loads(line) for line in f]
+    raw_ex_batch = raw_ex_batch[:num_prompts]
+    requests, sampling_params = example_to_requests(args, tokenizer, raw_ex_batch)
+
+    kwargs = {
+        "model": model_path,
+        "trust_remote_code": True,
+        "tensor_parallel_size": tp_size,
+        "dtype": torch.bfloat16,
+        "enforce_eager": enforce_eager,
+        "disable_custom_all_reduce": True,
+        "distributed_executor_backend": "ray",
```

Comment:
More configuration parameters can be exposed, such as multi step scheduler.

Reply:
done